core/include/io.h

```diff
@@ -273,4 +273,31 @@ static inline void io_clrsetbits8(vaddr_t addr, uint8_t clear_mask,
 	io_write8(addr, (io_read8(addr) & ~clear_mask) | set_mask);
 }
 
+/*
+ * Poll on a IO memory content or timeout
+ *
+ * @_addr is the address of the memory cell accessed
+ * @_val represents the val  of the memory cell accessed
+ * @_cond represents the condition to get the correct value
+ * @_delay_us represents the read interval in mircorseconds
+ * @_timeout_us represents the time-out period in microseconds
+ *
+ * @return nonzero value means time-out, 0 means got right value
+ */
+#define IO_READ32_POLL_TIMEOUT(_addr, _val, _cond, _delay_us, _timeout_us) \
+	({ \
+		uint64_t __timeout = 0; \
+		uint64_t __delay = (_delay_us); \
```

Comment:
Prefer `uint32_t`: `udelay()` argument is `uint32_t`.
`__timeout` also could be `uint32_t`.